File: src/com/tufar/IPCalculator/MainActivity.java
package com.tufar.IPCalculator;

import com.tufar.IPCalculator.R;


import android.app.Activity;
import android.os.Bundle;
import android.view.Menu;
import android.widget.NumberPicker;
import android.widget.NumberPicker.OnScrollListener;
import android.widget.TextView;

public class MainActivity extends Activity {
	NumberPicker ip1;
	NumberPicker ip2;
	NumberPicker ip3;
	NumberPicker ip4;
	NumberPicker netmask_one_number;
	TextView subnetMaskExpanded;
	TextView numberOfHosts;
	TextView broadcastAddress;
	TextView wildcardMask;
	TextView hostsAddressRange;
	TextView netmaskBinary;
	
	IPv4 ipv4 = new IPv4("10.1.0.25/16");
	
	
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        
        subnetMaskExpanded = (TextView)findViewById(R.id.subnetMaskExpanded);
        numberOfHosts = (TextView)findViewById(R.id.numberOfHosts);
        broadcastAddress = (TextView)findViewById(R.id.broadcastAddress);
        wildcardMask = (TextView)findViewById(R.id.wildcardMask);
        hostsAddressRange = (TextView)findViewById(R.id.hostsAddressRange);
        netmaskBinary = (TextView)findViewById(R.id.netmaskBinary);
        
        ip1 = (NumberPicker)findViewById(R.id.ip1);
        ip1.setMinValue(0);
        ip1.setMaxValue(255);
        ip1.setValue(192);
        ip1.setOnScrollListener(updateTheResults);
        ip1.setDescendantFocusability(NumberPicker.FOCUS_BLOCK_DESCENDANTS);

        ip2 = (NumberPicker)findViewById(R.id.ip2);
        ip2.setMinValue(0);
        ip2.setMaxValue(255);
        ip2.setValue(168);
        ip2.setOnScrollListener(updateTheResults);
        ip2.setDescendantFocusability(NumberPicker.FOCUS_BLOCK_DESCENDANTS);

        ip3 = (NumberPicker)findViewById(R.id.ip3);
        ip3.setMinValue(0);
        ip3.setMaxValue(255);
        ip3.setValue(0);
        ip3.setOnScrollListener(updateTheResults);
        ip3.setDescendantFocusability(NumberPicker.FOCUS_BLOCK_DESCENDANTS);

        ip4 = (NumberPicker)findViewById(R.id.ip4);
        ip4.setMinValue(0);
        ip4.setMaxValue(255);
        ip4.setValue(0);
        ip4.setOnScrollListener(updateTheResults);
        ip4.setDescendantFocusability(NumberPicker.FOCUS_BLOCK_DESCENDANTS);

        netmask_one_number = (NumberPicker)findViewById(R.id.netmask_one_number);
        netmask_one_number.setMinValue(8);
        netmask_one_number.setMaxValue(31);
        netmask_one_number.setValue(16);
        netmask_one_number.setOnScrollListener(updateTheResults);
        netmask_one_number.setDescendantFocusability(NumberPicker.FOCUS_BLOCK_DESCENDANTS);
    }
    
    OnScrollListener updateTheResults = new OnScrollListener() {
		@Override
		public void onScrollStateChange(NumberPicker view, int scrollState) {
			Integer ip1i = ip1.getValue();
			Integer ip2i = ip2.getValue();
			Integer ip3i = ip3.getValue();
			Integer ip4i = ip4.getValue();
			Integer netmask_one_number_i = netmask_one_number.getValue();
			
			String CIDR = 	ip1i.toString()+"."+
							ip2i.toString()+"."+
							ip3i.toString()+"."+
							ip4i.toString()+"/"+
							netmask_one_number_i.toString();
			try{
				ipv4 = new IPv4(CIDR);
				subnetMaskExpanded.setText(ipv4.getNetmask());
				numberOfHosts.setText(ipv4.getNumberOfHosts().toString());
				broadcastAddress.setText(ipv4.getBroadcastAddress());
				wildcardMask.setText(ipv4.getWildcardMask());
				hostsAddressRange.setText(ipv4.getHostAddressRange());
				netmaskBinary.setText(ipv4.getNetmaskInBinary());
			}catch(Exception e){
				subnetMaskExpanded.setText(e.getMessage());
			}
		}
	};
    
    


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    
}

File: app/src/main/java/com/tufar/IPCalculator/MainActivity.kt
package com.tufar.IPCalculator

import android.os.Bundle
import android.widget.NumberPicker
import androidx.appcompat.app.AppCompatActivity
import com.tufar.IPCalculator.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var ipv4 = IPv4("192.168.0.0/16")

    private val onValueChange = NumberPicker.OnValueChangeListener { _, _, _ ->
        calculateIp()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupNumberPicker(binding.ip1, 0, 255, 192)
        setupNumberPicker(binding.ip2, 0, 255, 168)
        setupNumberPicker(binding.ip3, 0, 255, 0)
        setupNumberPicker(binding.ip4, 0, 255, 0)
        setupNumberPicker(binding.netmaskOneNumber, 8, 31, 16)

        calculateIp()
    }

    private fun setupNumberPicker(picker: NumberPicker, min: Int, max: Int, value: Int) {
        picker.minValue = min
        picker.maxValue = max
        picker.value = value
        picker.setOnValueChangedListener(onValueChange)
        picker.descendantFocusability = NumberPicker.FOCUS_BLOCK_DESCENDANTS
    }

    private fun calculateIp() {
        val cidr = "${binding.ip1.value}.${binding.ip2.value}.${binding.ip3.value}.${binding.ip4.value}/${binding.netmaskOneNumber.value}"
        binding.networkDisplay.text = cidr
        try {
            ipv4 = IPv4(cidr)
            binding.subnetMaskExpanded.text = ipv4.netmask
            binding.numberOfHosts.text = ipv4.numberOfHosts.toString()
            binding.broadcastAddress.text = ipv4.broadcastAddress
            binding.wildcardMask.text = ipv4.wildcardMask
            binding.hostsAddressRange.text = ipv4.hostAddressRange
            binding.netmaskBinary.text = ipv4.netmaskInBinary
        } catch (e: Exception) {
            binding.subnetMaskExpanded.text = e.message
        }
    }
}
